File: 2015/py/src/day07.py
# Day 07: Some Assembly Required

from src import util

circuit: dict[str, list[str]] = {}
output: dict[str, int] = {}

instructions: list[str] = util.input_lines("day07.txt")
for i in instructions:
    lhs, rhs = i.split("->")
    rhs = rhs.strip()
    circuit[rhs] = lhs.split()


def solve(wire: str) -> int:
    if wire not in output:
        if wire.isnumeric():
            value = int(wire)
        else:
            i = circuit[wire]
            if len(i) == 1:
                value = solve(i[0])
            else:
                match i[1]:
                    case "LSHIFT":
                        value = solve(i[0]) << solve(i[2])
                    case "RSHIFT":
                        value = solve(i[0]) >> solve(i[2])
                    case "AND":
                        value = solve(i[0]) & solve(i[2])
                    case "OR":
                        value = solve(i[0]) | solve(i[2])
                    case _:
                        value = ~solve(i[1])

        output[wire] = value

    return output[wire]


def star1():
    return solve("a")


def star2():
    global output
    circuit["b"] = [str(output["a"])]

    output = {}
    return solve("a")

File: 2024/jl/day07.jl
# Day 7: Bridge Repair

include("util.jl")

using .Util

data = []
for i ∈ Util.getinput("day07.txt")
	s = split(i, ": ")
	push!(data, (parse(Int64, s[1]), parse.(Int64, split(s[2]))))
end

function star1()
	total = 0
	println("- $total")
end

function star2()
	total = 0
	println("- $total")
end

star1()
star2()

